File: src/main/java/com/connectflow/service/UserService.java
package com.connectflow.service;

import com.connectflow.dto.CreateUserRequest;
import com.connectflow.dto.PageResponse;
import com.connectflow.dto.UserDTO;
import com.connectflow.model.User;
import com.connectflow.model.UserRole;
import com.connectflow.repository.UserRepository;
import com.connectflow.repository.UserRoleRepository;
import com.connectflow.repository.BranchRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional
public class UserService {

    private final UserRepository userRepository;
    private final UserRoleRepository userRoleRepository;
    private final BranchRepository branchRepository;

    /**
     * Get user by ID
     */
    public Optional<UserDTO> getUserById(UUID id) {
        return userRepository.findById(id).map(this::convertToDTO);
    }

    /**
     * Get user by email
     */
    public Optional<UserDTO> getUserByEmail(String email) {
        return userRepository.findByEmail(email).map(this::convertToDTO);
    }

    /**
     * Get all users
     */
    public List<UserDTO> getAllUsers() {
        return userRepository.findAll().stream()
            .map(this::convertToDTO)
            .collect(Collectors.toList());
    }

    /**
     * Get users by role
     */
    public List<UserDTO> getUsersByRole(UserRole.Role role) {
        return userRoleRepository.findByRole(role).stream()
            .map(userRole -> userRepository.findById(userRole.getUserId()))
            .filter(Optional::isPresent)
            .map(Optional::get)
            .map(this::convertToDTO)
            .collect(Collectors.toList());
    }

    /**
     * Get users in a branch
     */
    public List<UserDTO> getUsersByBranch(UUID branchId) {
        return userRoleRepository.findByBranchId(branchId).stream()
            .map(userRole -> userRepository.findById(userRole.getUserId()))
            .filter(Optional::isPresent)
            .map(Optional::get)
            .map(this::convertToDTO)
            .collect(Collectors.toList());
    }

    /**
     * Get all users with pagination
     */
    public PageResponse<UserDTO> getAllUsersPaginated(int page, int size, String sortBy, String sortDir) {
        Sort sort = sortDir.equalsIgnoreCase("desc")
            ? Sort.by(sortBy).descending()
            : Sort.by(sortBy).ascending();

        Pageable pageable = PageRequest.of(page, size, sort);
        Page<User> userPage = userRepository.findAll(pageable);

        List<UserDTO> content = userPage.getContent().stream()
            .map(this::convertToDTO)
            .collect(Collectors.toList());

        return new PageResponse<>(
            content,
            userPage.getNumber(),
            userPage.getSize(),
            userPage.getTotalElements(),
            userPage.getTotalPages(),
            userPage.isLast()
        );
    }

    /**
     * Create a new user with role and branch assignment
     */
    public UserDTO createUser(CreateUserRequest request) {
        // Validate required fields
        if (request.getEmail() == null || request.getEmail().trim().isEmpty()) {
            throw new IllegalArgumentException("Email is required");
        }
        if (request.getPassword() == null || request.getPassword().trim().isEmpty()) {
            throw new IllegalArgumentException("Password is required");
        }
        if (request.getFullName() == null || request.getFullName().trim().isEmpty()) {
            throw new IllegalArgumentException("Full name is required");
        }
        if (request.getRole() == null) {
            throw new IllegalArgumentException("Role is required");
        }

        // Check if email already exists
        if (userRepository.findByEmail(request.getEmail().trim().toLowerCase()).isPresent()) {
            throw new IllegalArgumentException("Email already exists");
        }

        // Create the user
        User user = User.builder()
            .id(UUID.randomUUID())
            .fullName(request.getFullName().trim())
            .email(request.getEmail().trim().toLowerCase())
            .phone(request.getPhone())
            .password(request.getPassword().trim())
            .build();

        User savedUser = userRepository.save(user);

        // Create user role assignment
        UserRole userRole = UserRole.builder()
            .userId(savedUser.getId())
            .role(request.getRole())
            .branchId(request.getBranchId())
            .build();

        userRoleRepository.save(userRole);

        return convertToDTO(savedUser);
    }

    /**
     * Authenticate a user by email and password
     */
    public Optional<UserDTO> authenticate(String email, String password) {
        if (email == null || password == null) {
            return Optional.empty();
        }
        String normalizedEmail = email.trim().toLowerCase();
        String normalizedPassword = password.trim();

        return userRepository.findByEmail(normalizedEmail)
            .filter(user -> normalizedPassword.equals(user.getPassword()))
            .map(this::convertToDTO);
    }

    /**
     * Set PIN for a user (typically for branch managers)
     */
    public UserDTO setPin(UUID userId, String pin) {
        if (pin == null || pin.trim().isEmpty()) {
            throw new IllegalArgumentException("PIN cannot be empty");
        }
        if (!pin.matches("^\\d{4,6}$")) {
            throw new IllegalArgumentException("PIN must be 4-6 digits");
        }

        User user = userRepository.findById(userId)
            .orElseThrow(() -> new RuntimeException("User not found with id: " + userId));

        user.setPin(pin.trim());
        User updated = userRepository.save(user);
        return convertToDTO(updated);
    }

    /**
     * Verify PIN for a user
     */
    public boolean verifyPin(UUID userId, String pin) {
        if (pin == null || pin.trim().isEmpty()) {
            throw new IllegalArgumentException("PIN cannot be empty");
        }

        User user = userRepository.findById(userId)
            .orElseThrow(() -> new RuntimeException("User not found with id: " + userId));

        if (user.getPin() == null || user.getPin().trim().isEmpty()) {
            throw new RuntimeException("User does not have a PIN set");
        }

        return pin.trim().equals(user.getPin());
    }

    /**
     * Check if user has a PIN set
     */
    public boolean hasPinSet(UUID userId) {
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new RuntimeException("User not found with id: " + userId));

        return user.getPin() != null && !user.getPin().trim().isEmpty();
    }

    /**
     * Convert User entity to UserDTO
     */
    private UserDTO convertToDTO(User user) {
        // Get user's primary role
        Optional<UserRole> userRole = userRoleRepository.findByUserId(user.getId()).stream().findFirst();

        UserRole.Role role = null;
        UUID branchId = null;
        String branch = null;

        if (userRole.isPresent()) {
            role = userRole.get().getRole();
            if (userRole.get().getBranchId() != null) {
                branchId = userRole.get().getBranchId();
                branch = branchRepository.findById(branchId)
                    .map(com.connectflow.model.Branch::getName)
                    .orElse(null);
            }
        }

        return UserDTO.builder()
            .id(user.getId())
            .fullName(user.getFullName())
            .email(user.getEmail())
            .phone(user.getPhone())
            .role(role)
            .branchId(branchId)
            .branch(branch)
            .createdAt(user.getCreatedAt())
            .updatedAt(user.getUpdatedAt())
            .build();
    }
}

File: next/backend/src/service/UserService.js
'use strict';
const bcrypt = require('bcryptjs');
const { v4: uuidv4 } = require('uuid');
const UserRepository = require('../repository/UserRepository');
const UserRoleRepository = require('../repository/UserRoleRepository');
const { generateToken } = require('../security/JwtService');

module.exports = {
  async login(email, password) {
    const user = await UserRepository.findByEmail(email);
    if (!user) throw { status: 401, message: 'Invalid email or password' };
    const valid = await bcrypt.compare(password, user.password);
    if (password != user.password) throw { status: 401, message: 'Invalid email or password' };
    const token = generateToken(user.email);
    const primaryRole = (user.roles && user.roles[0]) || null;
    return {
      token,
      user: {
        id: user.id,
        fullName: user.fullName,
        email: user.email,
        phone: user.phone || null,
        role: primaryRole ? primaryRole.role : null,
        branchId: primaryRole ? primaryRole.branchId : null,
        branch: primaryRole && primaryRole.branch ? primaryRole.branch.name : null,
      },
    };
  },

  async getAllUsers() {
    return UserRepository.findAll();
  },

  async getUserById(id) {
    const user = await UserRepository.findById(id);
    if (!user) throw { status: 404, message: 'User not found' };
    return user;
  },

  async getUserByEmail(email) {
    const user = await UserRepository.findByEmail(email);
    if (!user) throw { status: 404, message: 'User not found' };
    return user;
  },

  async getUsersByRole(role) {
    return UserRepository.findByRole(role);
  },

  async getUsersByBranch(branchId) {
    return UserRepository.findByBranch(branchId);
  },

  async getPaginatedUsers({ page = 0, size = 10, sortBy = 'created_at', sortDir = 'desc', name, email, role, branchId }) {
    const { count, rows } = await UserRepository.findPaginated({ page, size, sortBy, sortDir, name, email, role, branchId });
    return {
      content: rows,
      pageNumber: page,
      pageSize: size,
      totalElements: count,
      totalPages: Math.ceil(count / size),
      last: (page + 1) * size >= count,
    };
  },

  async createUser({ fullName, email, phone, password, role, branchId }) {
    const existing = await UserRepository.findByEmail(email);
    if (existing) throw { status: 409, message: 'Email already in use' };
    const hash = await bcrypt.hash(password, 12);
    const userId = uuidv4();
    const user = await UserRepository.create({ id: userId, fullName, email, phone, password: hash });
    if (role) {
      await UserRoleRepository.create({ id: uuidv4(), userId: user.id, role, branchId: branchId || null });
    }
    return UserRepository.findById(user.id);
  },

  async updateUser(id, data) {
    const user = await UserRepository.findById(id);
    if (!user) throw { status: 404, message: 'User not found' };
    const update = {};
    if (data.fullName) update.fullName = data.fullName;
    if (data.phone !== undefined) update.phone = data.phone;
    if (data.password) update.password = await bcrypt.hash(data.password, 12);
    await UserRepository.update(id, update);
    if (data.role) {
      await UserRoleRepository.deleteByUserId(id);
      await UserRoleRepository.create({ id: uuidv4(), userId: id, role: data.role, branchId: data.branchId || null });
    }
    return UserRepository.findById(id);
  },

  async setPin(id, pin) {
    const user = await UserRepository.findById(id);
    if (!user) throw { status: 404, message: 'User not found' };
    const hashedPin = await bcrypt.hash(pin, 12);
    await UserRepository.update(id, { pin: hashedPin });
    return { message: 'PIN set successfully' };
  },

  async verifyPin(id, pin) {
    const user = await UserRepository.findById(id);
    if (!user) throw { status: 404, message: 'User not found' };
    if (!user.pin) return { verified: false };
    const match = await bcrypt.compare(pin, user.pin);
    return { verified: match };
  },
  async verifyManagerPin(email, pin) {
    const user = await UserRepository.findByEmail(email);
    if (!user) throw { status: 401, message: 'Invalid credentials' };
    const role = user.roles && user.roles[0] ? user.roles[0].role : null;
    if (!role || !['SUPERADMIN', 'ADMIN', 'MANAGER'].includes(role)) {
      throw { status: 403, message: 'User does not have manager-level access' };
    }
    if (!user.pin) throw { status: 400, message: 'Manager has no PIN set' };
    const match = await bcrypt.compare(pin, user.pin);
    if (!match) throw { status: 401, message: 'Invalid PIN' };
    return { verified: true, role, userId: user.id };
  },


  async hasPin(id) {
    const user = await UserRepository.findById(id);
    if (!user) throw { status: 404, message: 'User not found' };
    return { hasPin: !!user.pin };
  },
};
